sql/vector_mhnsw.cc

```diff
@@ -270,185 +174,261 @@ static bool select_neighbours(TABLE *source, TABLE *graph,
+  neighbor_array->num= new_neighbours.elements;
+
+  int i= 0;
+  for (const auto &node: new_neighbours)
   {
-    if (err != HA_ERR_KEY_NOT_FOUND)
-      return -1;
-    return 0;
+    if (node.get_ref_len() != ref_length)
+    {
+      // this should never happen
+      return true;
+    }
+    memcpy(neighbor_array->neigh_ref + i * ref_length, node.get_ref(),
+           node.get_ref_len());
+    i++;
   }
 
-  /*
-  if ((err= graph->file->ha_index_next_same(graph->record[0], key,
-                                            graph->key_info->key_length)))
-    return err == HA_ERR_END_OF_FILE ? 0 : -1;
-  */
-  return 1;
-}
-
-static inline bool store_link(TABLE *graph,
-                              size_t layer_number,
-                              const FVectorRef &src,
-                              const FVectorRef &dst)
-{
-  //TODO(cvicentiu) this causes 2 uneeded stores because field 0 and 1 almost
-  //never change.
-  //Or we could iterate and store them all src->dest first and then all dest->src
   graph->field[0]->store(layer_number);
   graph->field[1]->store_binary(
-    reinterpret_cast<const char *>(src.get_ref()), src.get_ref_len());
-  graph->field[2]->store_binary(
-    reinterpret_cast<const char *>(dst.get_ref()), dst.get_ref_len());
-  return graph->file->ha_write_row(graph->record[0]);
-}
-
-
-static bool write_connection_to_graph(TABLE *graph,
-                                      uchar *key,
-                                      size_t layer_number,
-                                      const FVectorRef &src,
-                                      const FVectorRef &dst)
-{
-  // TODO(cvicentiu) if store_link fails on the second write, we can end up
-  // with a corrupt index. Handle this and write undo logic.
-  int con_exists= connection_exists(graph, key, layer_number, src, dst);
-  if (con_exists)
-    return con_exists < 0;
-
-  return store_link(graph, layer_number, src, dst) ||
-         store_link(graph, layer_number, dst, src);
-}
+    reinterpret_cast<const char *>(source_node.get_ref()),
+    source_node.get_ref_len());
+  graph->field[2]->set_null();
 
-static bool insert_node_connections_on_layer(TABLE *graph,
-                                             size_t layer_number,
-                                             const FVector &source_node,
-                                             const List<FVector> &dest_nodes)
-{
   uchar *key= (uchar*)alloca(graph->key_info->key_length);
+  key_copy(key, graph->record[0], graph->key_info, graph->key_info->key_length);
 
-  if (dest_nodes.elements == 0)
+  int err= graph->file->ha_index_read_map(graph->record[1], key,
+                                           HA_WHOLE_KEY,
+                                           HA_READ_KEY_EXACT);
+
+  // no record
+  if (err)
   {
-    return store_link(graph, layer_number, source_node, source_node);
+    graph->field[2]->store_binary(
+      reinterpret_cast<const char *>(neighbor_array), totalSize);
```

Comment:
This is not correct. One can not just store "structs" on disk. Between compilers / architectures the structs might be padded differently. A struct is an in-memory representation that is not guaranteed to be stable on disk. For storing things persistently, one must:

1. Store things byte-by-byte, explicitly.
2. Ensure that the storage happens using fixed endinanness. There are `uintxkorr` functions that can e used.

Note:
This is likely a problem with my first implementation too, when reading floats from disk. We can not assume floats are stored in little endian format. If we are to run on a big-endian machine, the data would not be portable.

Reply:
This is definitely a great point Vicentiu! I did understand that's not a portable solution for final implementation. In this commit I was mainly to make a workable version to see if the solution makes sense. I should've made an inline comment for this TODO. Also one aspect I wasn't very familiar is the reference size. As you may noticed in the code, I assumed all references of the same table have the same size (in my test the ref_size seems to be 6 bytes). I can look deeper into understanding how the references operate and maybe a more efficient structure for storing the neighbours.

Currently in the code I'm using this structure to save the neigh_ref:
```
struct NeighborArray
{
  uint8 num;
  char neigh_ref[];
};
```
Do you think we need `uint8 num` here to validate the length of neighbours read from the second table? Or should simply get the size of neighbour list size by `neigh_ref/<ref_len got from source table info>` ?